### DISCUSSION TITLE
Is there a way to pass payload to delay event?

### DISCUSSION THREAD

**Discussion by dsngo** [statelyai/xstate#2018] [Q&A] [answered]
Current I am facing a situation where passing a function to an event payload and execute that function from delay state is the only possible way. A simple example is that I get an eval function from the API (which can change depending on certain conditions), then I would like to delay that for a certain amount. My attempt for this is: 
```js
// import send
// eval the function string, save it as `fn`, and delay value as delayV
send({type: "ADD_FN", payload: {fn_1: fn, delay: delayV}});
// ... machine
{
  states: {
    test_0: { on: {TEST: "test_1"} },
    test_1: {
      after: {
        API_DELAY: { target: "test_2", actions: (_c, e) => e.payload.fn() } // this event is payloadless, and currently I cannot find a way to add payload to it.
      }
    },
    test_2: {
      type: "final"
    }
  },
  on: {
    ADD_FN: {target: ".test_1"}
  }
}, {
  // other options
  delays: {
    API_DELAY: (_c, e) => e.payload.delay
  }
}
```
Why question is, is there a way for this to happen, or it is just not the right way for xstate?

**Comment by Andarist** [comment, marked as answer]
There is no way to add an additional payload to the auto-generated events for `after` transitions. Not though that those are really just a sugar syntax - you can achieve the same with regular delayed events and those would allow you to add a payload.

Take a look how this sugar gets handled internally:
https://github.com/davidkpiano/xstate/blob/f3870f837e7d86bd4e3a1d28485e76b026579291/packages/core/src/StateNode.ts#L589-L594

**Reply by dsngo** [reply]
This is great, I think looking at the doc is not enough for me now, thanks @Andarist .

**Reply by dsngo** [reply]
@Andarist I realize that xstate is a bit inconsistent on where to interpolate parameter as function or object. For example the send actions creator:
```js
send((c,e) => ({type: "EVENT", key: e.payload})) // this works
send(({type: "EVENT", key: (c,e) => e.payload})) // this works
send("EVENT", {delay: (c,e) => e.delay}) // this works
send("EVENT", (c,e) => ({delay: e.delay})) // this won't work
// ... a few more that I couldn't remember.
```
I've learnt to use this pattern from xstate and just love using it everywhere. However, do you have any particular reason why some of the parameters are used, and some are not?

**Reply by Andarist** [reply]
Those are not equivalent though and some of those that you have listed do different things.

> send((c,e) => ({type: "EVENT", key: e.payload}))

This creates a simple event object. Dynamic event names & payloads are allowed (as per SCXML)

> send(({type: "EVENT", key: (c,e) => e.payload})) 

This probably doesn't do what you think. This creates an event that holds a function - the `key` **is not** resolved.

> send("EVENT", {delay: (c,e) => e.delay})

The presence of `delay` is static, quite useful for visualization. It works - creates a related event object and it's an equivalent of `send({ type: "EVENT" }, {delay: (c,e) => e.delay})`

> send("EVENT", (c,e) => ({delay: e.delay}))

This is, in fact, not allowed.

The difference is this - the first argument describes the event's shape and can be dynamic. The second argument holds additional options for the dispatcher which describes when (`delay`) and where (`to`) the created event should be dispatched. For the second argument each of those "meta" properties can be dynamic, but their presence is required to be static.

**Reply by dsngo** [reply]
Very much what I was looking for, thank you so much @Andarist .
